### PR TITLE
upgrade to go1.12.1

### PR DESCRIPTION
**Pull request by BenTheElder**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**: 
- Updates the kube-cross image to install go 1.12.1 while we wait on https://github.com/docker-library/official-images/pull/5550
- Upgrades the bazel build to rules_go 0.17.2 and go1.12.1
- Upgrades test/images to go1.12.1

We need this so we can get more CI signal with go 1.12.1.

When to upstream golang 1.12.1 image is out we should update kube-cross to use that, in the meantime we've borrowed the go install snippet to install 1.12.1 on top of the 1.12 base image.

**Which issue(s) this PR fixes**:

Related to https://github.com/kubernetes/kubernetes/issues/75372

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Update to use go1.12.1
```

/sig release

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @dims @amwat 
need to actually build this next.

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/75372

**Comment by BenTheElder** [comment]
I just build the cross image locally and then `docker tag staging-k8s.gcr.io/kube-cross:v1.12.1-1 k8s.gcr.io/kube-cross:v1.12.1-1`, running a `make quick-release to sanity check it.

**Comment by BenTheElder** [comment]
`make quick-release` sanity checks fine, also:
```
$ build/run.sh go version
+++ [0315 11:08:17] Verifying Prerequisites....
+++ [0315 11:08:17] Building Docker image kube-build:build-f5f495327d-5-v1.12.1-1
+++ [0315 11:08:22] Keeping container kube-build-data-f5f495327d-5-v1.12.1-1
+++ [0315 11:08:22] Keeping container kind-control-plane
+++ [0315 11:08:22] Keeping container kube-build-data-14b730b626-5-v1.11.5-1
+++ [0315 11:08:22] Keeping container awesome_fermi
+++ [0315 11:08:22] Keeping container musing_shockley
+++ [0315 11:08:22] Keeping container laughing_hodgkin
+++ [0315 11:08:22] Keeping container kube-build-data-f5f495327d-5-v1.12.1-1
+++ [0315 11:08:22] Keeping container kind-control-plane
+++ [0315 11:08:22] Keeping container kube-build-data-14b730b626-5-v1.11.5-1
+++ [0315 11:08:22] Keeping container awesome_fermi
+++ [0315 11:08:22] Keeping container musing_shockley
+++ [0315 11:08:22] Keeping container laughing_hodgkin
+++ [0315 11:08:22] Keeping container kube-build-data-f5f495327d-5-v1.12.1-1
+++ [0315 11:08:22] Keeping container kind-control-plane
+++ [0315 11:08:22] Keeping container kube-build-data-14b730b626-5-v1.11.5-1
+++ [0315 11:08:22] Keeping container awesome_fermi
+++ [0315 11:08:22] Keeping container musing_shockley
+++ [0315 11:08:22] Keeping container laughing_hodgkin
+++ [0315 11:08:22] Keeping image kube-build:build-f5f495327d-5-v1.12.1-1
+++ [0315 11:08:22] Keeping image kube-build:build-14b730b626-5-v1.11.5-1
+++ [0315 11:08:22] Syncing sources to container
+++ [0315 11:08:22] Stopping any currently running rsyncd container
+++ [0315 11:08:22] Starting rsyncd container
+++ [0315 11:08:23] Running rsync
+++ [0315 11:08:24] Stopping any currently running rsyncd container
+++ [0315 11:08:25] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [0315 11:08:25] Running build command...
go version go1.12.1 linux/amd64
+++ [0315 11:08:27] Syncing out of container
+++ [0315 11:08:27] Stopping any currently running rsyncd container
+++ [0315 11:08:27] Starting rsyncd container
+++ [0315 11:08:28] Running rsync
+++ [0315 11:08:28] Stopping any currently running rsyncd container
```

**Comment by BenTheElder** [comment]
[filed a CL to promote the image, we are waiting on that now]

**Comment by BenTheElder** [comment]
/priority critical-urgent
/kind cleanup
(we really need a better set of `kind` to cover things like this)

**Comment by dims** [comment]
NOTE: we dont' yet have 1.12.1 golang image as of today (Fri Mar 15, approx 2:30 PM East coast). To make progress over the weekend to get good CI signal, we need this PR to merge quickly. 

Once we get an official image early next week, we can remove the hack in Dockerfile and use the 1.12.1 in the `FROM` directive.

Thanks for doing this @BenTheElder much appreciated

/approve
/lgtm 

(in principle)

File: build/build-image/cross/Dockerfile [hunk 1]
```diff
@@ -17,6 +17,50 @@
 
 FROM golang:1.12.0
```
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: the 1.12.1 tag is now published. do we still want to merge this workaround?

Comment (dims) [reply build/build-image/cross/Dockerfile]: @ixdy yes, This lets get the CI jobs get started quicker. We can revert things back on monday (and publish fresh images)

Comment (dims) [reply build/build-image/cross/Dockerfile]: https://github.com/kubernetes/kubernetes/issues/75372#issuecomment-473420258

Comment (ixdy) [reply build/build-image/cross/Dockerfile]: The PR merged a few hours ago, but you can also now pull the image, meaning this workaround isn't needed: 
```console
$ docker pull golang:1.12.1
1.12.1: Pulling from library/golang
22dbe790f715: Pull complete
0250231711a0: Pull complete
6fba9447437b: Pull complete
c2b4d327b352: Pull complete
619f4932b7ea: Pull complete
16e105463ffe: Pull complete
ed842bbe1793: Pull complete
Digest: sha256:8630e0f0e9a8187f54066ea81a9fe33a21ddbfbe88c7149fe2299d8cd989ad63
Status: Downloaded newer image for golang:1.12.1
```

Comment (dims) [reply build/build-image/cross/Dockerfile]: @ixdy i am happy for us to switch things back right now if you, @amwat and @BenTheElder are ok with it.

Comment (liggitt) [reply build/build-image/cross/Dockerfile]: build-cross v1.12.1-1 is already pushed with this workaround, so we should commit this to record the source of v1.12.1-1. that will also let CI soak begin for the weekend while a v1.12.1-2 without this workaround gets pushed.

Comment (BenTheElder) [reply build/build-image/cross/Dockerfile]: I'm up to switch it but it sounds like for now we want to let the CI soak then switch it for v1.12.1-2?
https://kubernetes.slack.com/archives/C2C40FMNF/p1552691032534800

**Comment by dims** [comment]
please see discussion at https://kubernetes.slack.com/archives/C2C40FMNF/p1552690316531000 to approve this PR

**Comment by liggitt** [comment]
/approve

xref 
https://github.com/kubernetes/kubernetes/pull/75413#discussion_r266172057 and 
https://kubernetes.slack.com/archives/C2C40FMNF/p1552691018534400

**Comment by BenTheElder** [comment]
That has to be the least useful junit output ever.

/retest

**Pull request merged**